src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -191,13 +221,189 @@ ErrorIfIllegalPartitioningInTenantSchema(Oid parentRelationId, Oid partitionRela
+ *  - Schema name is in the allowed-list,
+ *  - Schema is not owned by an extension,
+ *  - Current user should be the owner of tables under the schema,
+ *  - Table kinds are supported,
+ *  - Only Citus local and Postgres local tables exist under the schema.
+ */
+static void
+EnsureSchemaCanBeDistributed(Oid schemaId, List *schemaTableIdList)
+{
+	/* Ensure schema name is allowed */
+	EnsureTenantSchemaNameAllowed(schemaId);
+
+	/* Any schema owned by extension is not allowed */
+	ObjectAddress *schemaAddress = palloc0(sizeof(ObjectAddress));
+	ObjectAddressSet(*schemaAddress, NamespaceRelationId, schemaId);
+	if (IsAnyObjectAddressOwnedByExtension(list_make1(schemaAddress), NULL))
+	{
+		char *schemaName = get_namespace_name(schemaId);
+		ereport(ERROR, (errmsg("schema %s, which is owned by an extension, cannot "
+							   "be distributed", schemaName)));
+	}
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, schemaTableIdList)
+	{
+		/* Ensure table owner */
+		EnsureTableOwner(relationId);
+
+		/* Check relation kind */
+		EnsureTableKindSupportedForTenantSchema(relationId);
+
+		/* Postgres local tables are allowed */
+		if (!IsCitusTable(relationId))
+		{
+			continue;
+		}
+
+		/* Only Citus local tables, amongst Citus table types, are allowed */
+		if (!IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+		{
+			ereport(ERROR, (errmsg("schema already has distributed tables"),
+							errhint("Undistribute distributed tables under "
+									"the schema before distributing the schema.")));
+		}
+	}
+}
+
+
+/*
+ * EnsureTenantSchemaNameAllowed ensures if given schema is applicable for registering
+ * as a tenant schema.
+ */
+static void
+EnsureTenantSchemaNameAllowed(Oid schemaId)
+{
+	char *schemaName = get_namespace_name(schemaId);
+
+	/* public schema is not allowed */
+	if (strcmp(schemaName, "public") == 0)
+	{
+		ereport(ERROR, (errmsg("public schema cannot be distributed")));
+	}
+
+	/* information_schema schema is not allowed */
+	if (strcmp(schemaName, "information_schema") == 0)
+	{
+		ereport(ERROR, (errmsg("information_schema schema cannot be distributed")));
+	}
+
+	/* pg_temp_xx and pg_toast_temp_xx schemas are not allowed */
+	if (isAnyTempNamespace(schemaId))
+	{
+		ereport(ERROR, (errmsg("temporary schema cannot be distributed")));
+	}
+
+	/* pg_catalog schema is not allowed */
+	if (IsCatalogNamespace(schemaId))
+	{
+		ereport(ERROR, (errmsg("pg_catalog schema cannot be distributed")));
+	}
+
+	/* pg_toast schema is not allowed */
+	if (IsToastNamespace(schemaId))
+	{
+		ereport(ERROR, (errmsg("pg_toast schema cannot be distributed")));
+	}
+}
+
+
+/*
+ * EnsureSchemaExist ensures that schema exists. Caller is responsible to take
+ * the required lock on the schema.
+ */
+static void
+EnsureSchemaExist(Oid schemaId)
+{
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
```

Comment:
This causes a bit weird error message on weird input, can we use a more generic error message than this?

```sql
select citus_schema_undistribute(1);
ERROR:  concurrent operation removed the schema
```

Maybe "schema with OID %u does not exist"?